SaladTechnologies/oras-go
# [RD] Add Salad's resumable download patches

These commits add resumable download capability to oras-go's Copy() method.

- Annotations key declarations (internal/spec/artifact.go)
  - Constants for the Annotations map used by resumable downloads
- Copy hashVerifier from go-digest
  - hashVerifier is non-public in go-digest and we need to supply a pre-initialized Hash
- Add resume to VerifyReader and add hashVerifier (content/reader.go)
  - Add resume field to VerifyReader to track resume mode
  - Skip setting UnexpectedEOF error in Read() in resume mode
  - Add hashVerifier from digest
  - Recover serialized Hash object in NewVerifyReader and create a content.hashVerifier if in resume mode
  - Duplicate tests for resume mode
- Add Store.IngestFile() and Storage.IngestFile() (content/oci/storage.go)
  - injest(): calculate hash on existing ingest file contents
- Add Repository.FetchHead() and friends (registry/remote/repository.go)
  - BlobStoreHead.FetchHead()
  - Repository.FetchHead()
  - blobStore.Fetch() checks for Range header support and sets it if so
- Updates to CopyBuffer (internal/ioutil/io.go)
  - Handle partial copies in CopyBuffer()
  - Add tests (upstream TestUnwrapNopCloser(0 is failing, not fixed here)
- Set up resume in doCopyNode() (copy.go)
  - Look for a partial downloaded file
  - Get headers from src to verify Range: support
  - Set up resume arguments in desc.Annotations
